src/mpi/attr/attrutil.c

```diff
@@ -97,11 +96,8 @@ int MPIR_Call_attr_delete(int handle, MPIR_Attribute * attr_p)
          * user expectation is probably that the user-provided error code
          * is returned. */
         mpi_errno = rc;
-#else
-        mpi_errno =
-            MPIR_Err_create_code(MPI_SUCCESS, MPIR_ERR_RECOVERABLE, __func__, __LINE__,
-                                 MPI_ERR_OTHER, "**user", "**userdel %d", rc);
 #endif
+        MPIR_ERR_SET1(mpi_errno, MPI_ERR_OTHER, "**user", "**userdel %d", rc);
```

Comment:
Looks like you always want to set a new error code here. I thought you wanted to pass through the user `rc`. What did I miss?

Reply:
The line right above the removed lines sets `mpi_errno = rc`. And then the next line is `MPIR_ERR_SET1(mpi_errno, ...`. Therefore the error do pass trough.

Reply:
`MPIR_ERR_SET1` creates a new error code, the old error code is stored in the error ring so we can show error stack on abort, but users won't be able to access that.

Reply:
I'd prefer to wait for #6950 to be merged to update this one and keep discussing about it. I tested this PR extensively with mpi4py, error propagation was definitely working fine. Unfortunately, all that testing is no longer relevant, as I had to completely change the way mpi4py handles errors in callbacks to not rely on MPI passing errors through.

Reply:
> `MPIR_ERR_SET1` creates a new error code, the old error code is stored in the error ring so we can show error stack on abort, but users won't be able to access that.

Sorry, Hui, it is complicated, but to my defense, this is not my code. You are not seeing the full picture of the error handling implementation in MPICH. The `MPIR_*` routines return some error code that is generated by using the various error macros like `MPIR_ERR_SET1`. You are right, this macro creates a new error code. The user error code get's eventually stored in the error ring and printed to the string, but the error ring definitely preserves the value. In the `MPI_*` layer, if the `MPIR_*` layer returns with error, then MPI error handling takes over, and eventually, the user error code is extracted from the ring and that's what the `MPI_*` caller gets.

It took me quite a while to figure out how all this machinery worked, fixing and consolidating stuff in a way that is easy to review is really hard, and that's the reason I imploring you to relax your rules for contribution in this particular instance. Anyway, I'm working right now on splitting one of my commits. Let's see how it goes.

Reply:
> I imploring you to relax your rules for contribution in this particular instance. Anyway, I'm working right now on splitting one of my commits. Let's see how it goes.

It's not a strict rule, but a guideline. I literally not getting the points of your fix and cannot finish the review or even to ask the right question, partly due to the commits are not clean. Splitting the commits allow reviewer to ignore the non-related changes and zoom in the the key contribution of the PR.

Reply:
> partly due to the commits are not clean. 

After building on top of your commit, I did some refactoring of my commits. Now there is one commit exclusively dedicated to to style/comments/whitespaces that do not change behavior at all, but just tries to consolidate things and match code in the implementations of comm/type/win. The really important commits are 82f5d69977e9b119c28f62baaead7ac677b8514b and  87a3f4750a397bce19225863449208a91ee02223, all the other are mostly preparatory cleanups.  Commit 4b9959ff2b6f8600e641b01bce0861e1809c9bcc  is admittedly involved, but it is what it takes to implement what the commit message describes.

All these commits are IMHO very clean and self contained. You may still not understand them, but it is really because of faulty style on my side, or just because the thing is really complicated? At this point I've spent hours looking at how all the pieces glue together, and there is no way you can grasp what's going on by just looking at the diff. How are we going to get out of this stagnation point?

Perhaps we should open a new PR with the these commits to start fresh and discard all the mess of intermixed comments, at this point it is very confusing. It would also be very constructive if you take a look at my comments below https://github.com/pmodels/mpich/pull/6935#issuecomment-2022721045 and tell me if you agree at least with the spirit of my intentions, otherwise there is no point in moving forward. Then maybe we can negotiate a more complete test case that exercises the behavior I described in full, such that a debugger can walk you through the mess and you can ultimately convince yourself that the implementation is sound.